### PR TITLE
NIF pids monitoring (shared slave)

### LINKED ISSUE #53: dpi:unload/0 should stop the slave only if its the last process using the slave node
With the introduction of configurable node_name for the dpi_slave we cannot call `dpi:unload/0` if there are multiple processes using the same slave.

### PR DESCRIPTION
**Pull request by c-bik**
fixes #53

### CODE REVIEW

File: src/dpi.erl [hunk 1]
```diff
@@ -66,11 +71,31 @@ load_unsafe() ->
...
+
+-spec(pids_set([pid()]) -> ok).
+pids_set(Pids) when is_list(Pids)  ->
+    exit({nif_library_not_loaded, dpi, pids_set}).
+
+register_process(Pid) ->
+    pids_set(
+        lists:filter(
+            fun(P) ->
+                true == (catch rpc:call(
```
Comment (shamis) [inline src/dpi.erl]: can be replaced with `rpc_call(node(P), erlang, is_process_alive, [P])` so the catch can be removed as well

File: src/dpi.erl [hunk 2]
```diff
@@ -51,13 +50,19 @@ load(SlaveNodeName) when is_atom(SlaveNodeName) ->
 
 unload() ->
     SlaveNode = erase(dpi_node),
-    slave:stop(SlaveNode).
+    Self = self(),
+    case catch rpc_call(SlaveNode, dpi, pids_get, []) of
+        [] -> slave:stop(SlaveNode);
+        [Self] -> slave:stop(SlaveNode);
+        Refs -> io:format("~p still referencing ~p~n", [Refs, SlaveNode])
```
Comment (shamis) [inline src/dpi.erl]: I think the rpc_call can return error tuple as well.
```erlang
    Refs when is_list(Refs) -> io:format("~p still referencing ~p~n", [Refs, SlaveNode]);
    _Error -> no_op
end.
```

File: src/dpi.erl [hunk 3]
```diff
@@ -85,6 +100,13 @@ load_unsafe() ->
 %   local helper functions
 %===============================================================================
 
+reg(SlaveNode, Pid) ->
+    Name = {?MODULE, SlaveNode, make_ref()},
+    case global:register_name(Name, Pid) of
+        yes -> SlaveNode;
+        no -> {error, "failed to register process globally"}
```
Comment (shamis) [inline src/dpi.erl]: I think if the reg fails then the Slave node have to be stopped, as from the caller point of view the load call failed.

File: src/dpi.erl [hunk 4]
```diff
@@ -85,6 +100,13 @@ load_unsafe() ->
 %   local helper functions
 %===============================================================================
 
+reg(SlaveNode, Pid) ->
+    Name = {?MODULE, SlaveNode, make_ref()},
```
Comment (shamis) [inline src/dpi.erl]: Adding `node()` to the Name will give information from which this reference was created.
`Name = {?MODULE, SlaveNode, node(), make_ref()},`
This should make is easier to monitor the distribution of the slave nodes

File: src/dpi.erl [hunk 5]
```diff
@@ -18,37 +18,52 @@
...
-                    case slave_call(SlaveNode, code, add_paths, [code:get_path()]) of
+                    case slave_call(
+                        SlaveNode, code, add_paths, [code:get_path()]
+                    ) of
                         ok ->
                             case slave_call(SlaveNode, dpi, load_unsafe, []) of
-                                ok -> SlaveNode;
-                                Error -> Error
+                                ok -> reg(SlaveNode, self());
+                                {error, _} = Error -> Error
```
Comment (shamis) [inline src/dpi.erl]: `slave_call/4` can return error in different formats I think

File: src/dpi.erl (hunk 3 above)
Comment (c-bik) [reply src/dpi.erl]: Done in `load/1` instead, as `{error, {already_running, ...}}` case from `load` should not stop the slave node

Comment (c-bik) [reply src/dpi.erl]: Done in `load/1` instead, as `{error, {already_running, ...}}` case from `load` should not stop the slave node

File: c_src/dpi_nif.c [hunk 6]
```diff
@@ -165,15 +159,17 @@ static int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info)
...
+    ErlNifEnv *env, void **priv_data, void **old_priv_data,
+    ERL_NIF_TERM load_info)
 {
     CALL_TRACE;
 
     oranif_st *st = enif_alloc(sizeof(oranif_st));
     if (st == NULL)
     {
-        E("failed allocate private structure of %zd bytes\r\n", sizeof(oranif_st));
+        E("failed allocate private structure of %zd bytes\r\n",
```
Comment (KarlKeiser) [inline c_src/dpi_nif.c]: should be `%zu`, see [here](https://stackoverflow.com/a/2524675/10411602)

**Pull request merged**